swarm/storage/pyramid.go

```diff
@@ -168,227 +229,203 @@ func (self *PyramidChunker) decrementWorkerCount() {
 		depth++
 	}
 	log.Trace("pyramid.chunker", "depth", depth)
 
 	// Add the root chunk entry
-	branchCount := int64(len(chunk.SData)-8) / self.hashSize
+	branchCount := int64(len(chunkData)-8) / self.hashSize
 	newEntry := &TreeEntry{
 		level:         depth - 1,
 		branchCount:   branchCount,
-		subtreeSize:   uint64(chunk.Size),
-		chunk:         chunk.SData,
-		key:           key,
+		subtreeSize:   uint64(chunkSize),
+		chunk:         chunkData,
+		key:           self.key,
 		index:         0,
 		updatePending: true,
 	}
-	chunkLevel[depth-1] = append(chunkLevel[depth-1], newEntry)
+	self.chunkLevel[depth-1] = append(self.chunkLevel[depth-1], newEntry)
 
 	// Add the rest of the tree
 	for lvl := depth - 1; lvl >= 1; lvl-- {
 
 		//TODO(jmozah): instead of loading finished branches and then trim in the end,
 		//avoid loading them in the first place
-		for _, ent := range chunkLevel[lvl] {
+		for _, ent := range self.chunkLevel[lvl] {
 			branchCount = int64(len(ent.chunk)-8) / self.hashSize
 			for i := int64(0); i < branchCount; i++ {
 				key := ent.chunk[8+(i*self.hashSize) : 8+((i+1)*self.hashSize)]
-				newChunk := retrieve(key, chunkC, quitC)
-				if newChunk == nil {
+				newChunkData, err := self.getter.Get(Reference(key))
+				if err != nil {
 					return errLoadingTreeChunk
 				}
-				bewBranchCount := int64(len(newChunk.SData)-8) / self.hashSize
+				newChunkSize := newChunkData.Size()
+				bewBranchCount := int64(len(newChunkData)-8) / self.hashSize
 				newEntry := &TreeEntry{
 					level:         lvl - 1,
 					branchCount:   bewBranchCount,
-					subtreeSize:   uint64(newChunk.Size),
-					chunk:         newChunk.SData,
+					subtreeSize:   uint64(newChunkSize),
+					chunk:         newChunkData,
 					key:           key,
 					index:         0,
 					updatePending: true,
 				}
-				chunkLevel[lvl-1] = append(chunkLevel[lvl-1], newEntry)
+				self.chunkLevel[lvl-1] = append(self.chunkLevel[lvl-1], newEntry)
 
 			}
 
 			// We need to get only the right most unfinished branch.. so trim all finished branches
-			if int64(len(chunkLevel[lvl-1])) >= self.branches {
-				chunkLevel[lvl-1] = nil
+			if int64(len(self.chunkLevel[lvl-1])) >= self.branches {
+				self.chunkLevel[lvl-1] = nil
 			}
 		}
 	}
 
 	return nil
 }
 
-func (self *PyramidChunker) prepareChunks(isAppend bool, chunkLevel [][]*TreeEntry, data io.Reader, rootKey []byte, quitC chan bool, wg *sync.WaitGroup, jobC chan *chunkJob, chunkC chan *Chunk, errC chan error, storageWG *sync.WaitGroup) {
+func (self *PyramidChunker) prepareChunks(isAppend bool) {
 	log.Debug("pyramid.chunker: prepareChunks", "isAppend", isAppend)
-	defer wg.Done()
+	defer self.wg.Done()
 
 	chunkWG := &sync.WaitGroup{}
 
 	self.incrementWorkerCount()
 
-	go self.processor(self.workerCount, jobC, chunkC, errC, quitC, storageWG)
+	go self.processor(self.workerCount)
 
 	parent := NewTreeEntry(self)
-	var unfinishedChunk *Chunk
+	var unfinishedChunk ChunkData
```

Comment:
Should this be `unfinishedChunkData` for clarity?

Reply:
@nolash Fixed in 3358a59371fa05a4be83ff080df131b5534096b4

Reply:
@nolash Fixed in 484e22c94864832b528b8506292f977603276a6a